src/main/flight/imu.c

```diff
@@ -643,55 +807,46 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
         static bool gpsHeadingInitialized = false;  // TODO - remove
         if (gpsHeadingInitialized) {
-            float groundspeedGain;  // IMU yaw gain to be applied in imuMahonyAHRSupdate from ground course,
-            if (FLIGHT_MODE(GPS_RESCUE_MODE)) {
-                // GPS_Rescue adjusts groundspeedGain during a rescue in a range 0 - 4.5,
-                //   depending on GPS Rescue state and groundspeed relative to speed to home.
-                groundspeedGain = gpsRescueGetImuYawCogGain();
-            } else {
-                // 0.0 - 10.0, heuristic based on GPS speed and stick state
-                groundspeedGain = imuCalcGroundspeedGain(dt);
-            }
-            DEBUG_SET(DEBUG_ATTITUDE, 2, lrintf(groundspeedGain * 100.0f));
-            float courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
-            cogErr = imuCalcCourseErr(courseOverGround) * groundspeedGain;
+            imuCalcCourseErr(&ahrsState, &imuRuntimeConfig, dt, gpsSol.groundSpeed, &rMat);
+            ahrsState.headingGain *= imuCalcGroundspeedGain(&stickSuppression, dt);
         } else if (gpsSol.groundSpeed > GPS_COG_MIN_GROUNDSPEED) {
             // Reset the reference and reinitialize quaternion factors when GPS groundspeed > GPS_COG_MIN_GROUNDSPEED
-            imuComputeQuaternionFromRPY(&qP, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
+            imuComputeQuaternionFromRPY(&ahrsState.nominal.attitude, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
+            imuComputeRotationMatrix(&rMat, &ahrsState.nominal.attitude);
+            imuUpdateEulerAngles(&rMat);
             gpsHeadingInitialized = true;
         }
     }
 #else
     UNUSED(useMag);
 #endif
 
+    vector3_t acc_bf;
+    vector3Scale(&acc_bf, &acc.accADC, acc.dev.acc_1G_rec);
 
-    float gyroAverage[XYZ_AXIS_COUNT];
-    for (int axis = 0; axis < XYZ_AXIS_COUNT; ++axis) {
-        gyroAverage[axis] = gyroGetFilteredDownsampled(axis);
-    }
+    imuCalcAccError(&ahrsState, &imuRuntimeConfig, dt, &gyroAverage, acc_bf, &rMat);
 
-    const bool useAcc = imuIsAccelerometerHealthy(); // all smoothed accADC values are within 10% of 1G
-    imuMahonyAHRSupdate(dt,
-                        DEGREES_TO_RADIANS(gyroAverage[X]), DEGREES_TO_RADIANS(gyroAverage[Y]), DEGREES_TO_RADIANS(gyroAverage[Z]),
-                        useAcc, acc.accADC.x, acc.accADC.y, acc.accADC.z,
-                        magErr, cogErr,
-                        imuCalcKpGain(currentTimeUs, useAcc, gyroAverage));
+    imuAhrsUpdate(&ahrsState, &imuRuntimeConfig, dt,
+                  &gyroAverage,
+                  dtSaturated, &rMat);
+
+    // Pre-compute rotation matrix from quaternion
+    imuComputeRotationMatrix(&rMat, &ahrsState.nominal.attitude);
+    imuUpdateEulerAngles(&rMat);
 
-    imuUpdateEulerAngles();
+    attitudeIsEstablished = true;  // TODO conditional on RP covariance
```

Comment:
(relevant code should probably test covariance directly to see if estimate is good enough)

Reply:
I mostly agree, but I'm not sure what would happen if is isUpright (which uses attitudeIsEstablished) went false during flight. Things would probably be fine, but it will take a lot of checking to be sure.